workflow/sync/sync_manager_test.go

```diff
@@ -654,13 +647,13 @@ func TestMutexWfLevel(t *testing.T) {
 		assert.False(t, status)
 		assert.True(t, wfUpdate)
 
-		mutex := concurrenyMgr.syncLockMap["default/Mutex/my-mutex"].(*PriorityMutex)
+		mutex := concurrenyMgr.syncLockMap["default/Mutex/my-mutex"].(*PrioritySemaphore)
```

Comment:
The point of the change is to remove the PriorityMutex type. It's an implementation detail not designed to be exposed (it shouldn't be Capitalised exported which isn't fixed here).

This is only in a test, and this unchecked type assertion is just to allow the test framework to inspect the internals of the structure. 

No actual code should ever care about the implementation, they should be using the `Semaphore` interface only.

Keeping it around like you suggest just adds unnecessary cognitive load in my opinion.

Reply:
> Keeping it around like you suggest just adds unnecessary cognitive load in my opinion.

Ironically enough, I was thinking the opposite 😅; if I were to read the type and see "semaphore" when I am working on a mutex, that might be confusing and so I'd have to delve in to realize it is a semaphore of 1 with no backing ConfigMap. Remembering that becomes cognitive load as well. The ConfigMap piece is a decently sized semantic difference between the two (although not part of the type directly) with some ramifications (e.g. #11890)

Even if the type isn't exported, one may very well look at the type of a variable to understand its semantics.

Reply:
I'm assuming if you're in this code you understand that a Mutex may be implemented as a simplified or specialized semaphore. This gives you the visibility that we're just using a semaphore to implement it, instead of needing to inspect it to find this out.

I'd considered doing what you suggested when creating this PR (and it did go through that as a form), but I preferred this.

Reply:
Yea there is cognitive load either way.

If you've considered and even tried that though and decided this was the better option, that works for me, thanks for elaborating on that.